File: ruby/insecure_deserialization/session_handling.rb
require 'sinatra'
require 'yaml'
require 'base64'
require 'json'

# VULN 1: YAML.load on user input - executes arbitrary Ruby objects via Psych deserialization (RCE in Ruby < 3.1)
post '/session/restore' do
  user_prefs = params[:preferences]
  preferences = YAML.load(user_prefs)
  session[:user_preferences] = preferences
  { status: 'ok' }.to_json
end

# VULN 2: eval() on JSON string from user input - arbitrary code execution
post '/session/apply_discount' do
  discount_rule = params[:rule]
  json_string = request.body.read
  result = eval(json_string)
  session[:discount] = result[:discount_pct]
  { applied: true }.to_json
end

# VULN 3: Marshal.load on base64-decoded session param - RCE via crafted session payload
get '/session/load' do
  session_param = params[:session_data]
  session_obj = Marshal.load(Base64.decode64(session_param))
  session[:user_id]   = session_obj[:user_id]
  session[:cart_id]   = session_obj[:cart_id]
  session[:role]      = session_obj[:role]
  redirect '/dashboard'
end

File: typescript/insecure_deserialization/sessionHandling.ts
import { Request, Response } from 'express';
import yaml from 'js-yaml';

interface SessionData {
    userId: number;
    role: string;
    cart: string[];
    preferences: Record<string, unknown>;
}

// VULN 1: js-yaml yaml.load() on user input without SafeLoader — enables arbitrary code execution via !!js/function tags
export function importUserPreferences(req: Request, res: Response): void {
    const { yamlPayload } = req.body as { yamlPayload: string };
    // yaml.load() (NOT yaml.safeLoad / FAILSAFE_SCHEMA) deserializes JS objects and functions
    const preferences = yaml.load(yamlPayload) as Record<string, unknown>;
    res.json({ message: 'Preferences imported', preferences });
}

export function parseShippingRules(req: Request, res: Response): void {
    const { rulesYaml } = req.body as { rulesYaml: string };
    // Attacker can craft !!js/function YAML tags to execute arbitrary code at parse time
    const shippingRules = yaml.load(rulesYaml);
    res.json({ shippingRules });
}

// VULN 2: eval() on JSON from request body — evaluates attacker-controlled JavaScript
export function evaluateCartRule(req: Request, res: Response): void {
    const { ruleExpression } = req.body as { ruleExpression: string };
    const parsed = JSON.parse(ruleExpression);
    // Intended to compute a cart discount expression, but eval executes arbitrary code
    const result = eval(parsed.expression);
    res.json({ result });
}

// VULN 3: new Function(userCode)() pattern — constructs and immediately invokes arbitrary function from user input
export function executeShippingPolicy(req: Request, res: Response): void {
    const { policyCode } = req.body as { policyCode: string };
    const sessionData: SessionData = req.session as unknown as SessionData;
    // Allows attacker to supply arbitrary function body that runs with access to session context
    const policyResult = new Function('session', 'order', policyCode)(sessionData, req.body.order);
    res.json({ policyResult });
}
